flink/src/main/java/org/apache/iceberg/flink/FlinkCatalog.java

```diff
@@ -619,7 +625,23 @@ public void alterPartitionColumnStatistics(ObjectPath tablePath, CatalogPartitio
   @Override
   public List<CatalogPartitionSpec> listPartitions(ObjectPath tablePath)
       throws CatalogException {
-    throw new UnsupportedOperationException();
+    Table table = icebergCatalog.loadTable(toIdentifier(tablePath));
+    CloseableIterable<FileScanTask> tasks = table.newScan().planFiles();
+    List<DataFile> dataFiles = Lists.newArrayList(CloseableIterable.transform(tasks, FileScanTask::file));
+
+    Set<CatalogPartitionSpec> set = Sets.newHashSet();
+    for (DataFile dataFile : dataFiles) {
+      Map<String, String> map = new HashMap<>();
+      PartitionData partitionData = (PartitionData) dataFile.partition();
+      Types.StructType structType = partitionData.getPartitionType();
+      for (int i = 0; i < partitionData.size(); i++) {
+        map.put(structType.fields().get(i).name(), partitionData.get(i).toString());
```

Comment:
To get data from `StructLike`, this can call `String.valueOf(partition.get(i, Object.class))`.

Partition values can be null, so using `String.valueOf` will avoid `NullPointerException`.

Reply:
You can get the partition type for each file using `DataFile.specId()` and looking up the spec in the table.